k2/csrc/context.cu

```diff
@@ -28,4 +28,52 @@ RegionPtr NewRegion(ContextPtr &context, std::size_t num_bytes) {
   return ans;
 }
 
+ParallelRunner::ParallelRunner(ContextPtr c) : c_(c) {
+  if (c_->GetDeviceType() == kCuda) {
+    auto ret = cudaEventCreate(&event_);
+    K2_CHECK_CUDA_ERROR(ret);
+    // record event on `c_->GetCudaStream` and will be waited on `NewStream`
+    ret = cudaEventRecord(event_, c_->GetCudaStream());
+    K2_CHECK_CUDA_ERROR(ret);
+  }
+}
+cudaStream_t ParallelRunner::NewStream() {
+  DeviceType d = c_->GetDeviceType();
+  if (d == kCpu) {
+    return kCudaStreamInvalid;
+  } else {
+    K2_CHECK_EQ(d, kCuda);
+    cudaStream_t stream;
+    auto ret = cudaStreamCreate(&stream);
+    K2_CHECK_CUDA_ERROR(ret);
+    streams_.push_back(stream);
+
+    ret = cudaStreamWaitEvent(stream, event_, 0);
```

Comment:
@danpovey I'm still doubting this: as we only record event for `c_>GetCudaStream` once in the construction, does't only the first call `NewStream` wait the event? 
But If we record event every time in `NewStream`, as `With` will override the default stream of `c`, we may make those streams waiting one another, so we should not do this, am I right?

Reply:
cudaStreamWaitEvent causes stream `stream` to block until even `event` has finished.
We don't cause the default stream to wait for the streams until the object's destructor (or its
Finish() call, which should do what the destructor would normally do), so it's doing the right thing.

Reply:
mm. What I was asking is about `event_` created on`c_->GetCudaStream` in the constructor (not those events created in the destructor, I know that's right.)

Anyway, forgot that, I just realized that we need to wait on `event_` in every streams (not only int the first call) created in `NewStream()` as they run asynchronously and we don't know which one will set the value of `event_` first.